docs/developers/developer-reference/rust-testing-framework.md

```diff
@@ -0,0 +1,456 @@
+    b_wrapper.add_mandos_sc_call(sc_call, Some(expect));
+
+    cf_setup
+        .blockchain_wrapper
+        .write_mandos_output("_generated_fund.scen.json");
+```
+
+You have to add this at the end of your `fund_test`. The more complex the call, the more arguments you'll have to add and such. The `SCCallMandos` struct has the `add_argument` method so you don't have to do any encoding by yourself.  
+
+## Testing queries
+
+Testing queries is similar to testing transactions, just with less arguments (since there is no caller, and no payment, and any modifications are automatically reverted): 
+
+```rust
+#[test]
+fn status_test() {
+    let mut cf_setup = setup_crowdfunding(crowdfunding_esdt::contract_obj);
+    let b_wrapper = &mut cf_setup.blockchain_wrapper;
+
+    b_wrapper.execute_query(&cf_setup.cf_wrapper, |sc| {
+        let status = sc.status();
+        assert_eq!(status, Status::FundingPeriod);
+    });
+
+    let sc_query = ScQueryMandos::new(cf_setup.cf_wrapper.address_ref(), "status");
+    let mut expect = TxExpectMandos::new(0);
+    expect.add_out_value(&Status::FundingPeriod);
+
+    b_wrapper.add_mandos_sc_query(sc_query, Some(expect));
+
+    cf_setup
+        .blockchain_wrapper
+        .write_mandos_output("_generated_query_status.scen.json");
+}
+```
+
+## Testing smart contract errors
+
+In the previous transaction test, we've tested the happy flow. Now let's see how we can check errors:
+
+```rust
+#[test]
+fn test_sc_error() {
+    let mut cf_setup = setup_crowdfunding(crowdfunding_esdt::contract_obj);
+    let b_wrapper = &mut cf_setup.blockchain_wrapper;
+    let user_addr = &cf_setup.first_user_address;
+
+    b_wrapper.set_egld_balance(user_addr, &rust_biguint!(1_000));
+
+    b_wrapper.execute_tx(
+        user_addr,
+        &cf_setup.cf_wrapper,
+        &rust_biguint!(1_000),
+        |sc| {
+            let result = sc.fund(managed_token_id!(b""), managed_biguint!(1_000));
+            assert_eq!(result, sc_error!("wrong token"));
+
+            let user_deposit = sc.deposit(&managed_address!(user_addr)).get();
+            let expected_deposit = managed_biguint!(0);
+            assert_eq!(user_deposit, expected_deposit);
+
+            StateChange::Revert
+        },
+    );
+
+    let mut sc_call = ScCallMandos::new(user_addr, cf_setup.cf_wrapper.address_ref(), "fund");
+    sc_call.add_egld_value(&rust_biguint!(1_000));
+
+    let mut expect = TxExpectMandos::new(4);
+    expect.set_message("wrong token");
+
+    b_wrapper.add_mandos_sc_call(sc_call, Some(expect));
+
+    cf_setup
+        .blockchain_wrapper
+        .write_mandos_output("_generated_sc_err.scen.json");
+}
+```
+
+Notice how we've changed the payment intentionally to an invalid token to check the error case. Also, we've changed the expected deposit to "0" instead of the previous "1_000". And lastly, the most important thing: the `StateChange::Revert` return result, so we don't commit any partial changes.  
```

Comment:
```suggestion
Notice how we've changed the payment intentionally to an invalid token to check the error case. Also, we've changed the expected deposit to "0" instead of the previous "1_000". And lastly, the most important thing: the `StateChange::Revert` return result, so we don't commit any partial changes to the blockchain wrapper.
```